flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/stream/MultiJoinTestPrograms.java

```diff
@@ -0,0 +1,461 @@
+                                    .build())
+                    .setupTableSource(
+                            SourceTestStep.newBuilder("Shipments")
+                                    .addSchema("user_id_3 STRING", "location STRING")
+                                    .addOption("changelog-mode", "I,UA,UB,D")
+                                    .producedValues(
+                                            Row.ofKind(RowKind.INSERT, "1", "London"),
+                                            Row.ofKind(RowKind.INSERT, "2", "Paris"),
+                                            Row.ofKind(RowKind.INSERT, "3", "New York"),
+                                            Row.ofKind(RowKind.INSERT, "9", "Melbourne"),
+                                            Row.ofKind(RowKind.INSERT, "10", "Lost Shipment"),
+                                            Row.ofKind(RowKind.INSERT, "3", "Brasília"))
+                                    .build())
+                    .setupTableSink(
+                            SinkTestStep.newBuilder("sink")
+                                    .addSchema(
+                                            "user_id STRING",
+                                            "name STRING",
+                                            "order_id STRING",
+                                            "payment_id STRING",
+                                            "location STRING")
+                                    .addOption("changelog-mode", "I,UA,UB,D")
+                                    .consumedValues(
+                                            "+I[1, Gus, order0, payment1, London]",
+                                            "+I[1, Gus, order1, payment1, London]",
+                                            // NO I[2, Bob..] here because Bob was deleted
+                                            "+I[3, Nomad, order3, payment3, New York]",
+                                            // NO I[4, David..] here because David has only 5
+                                            // dollars, so  this is not true u.cash >= p.price
+                                            "+I[5, Eve, null, payment5, null]", // Here because
+                                            // payment price as
+                                            // -1
+                                            // NO I[7, Welcher..] here because Welcher has no
+                                            // payment
+                                            "+I[6, Frank, order6, payment6, null]",
+                                            // NO I[8, Welcher..] here because there is no order and
+                                            // no payment
+                                            "+I[9, Charlie Taylor, order9, payment9, Melbourne]", // Charlie Taylor because the name was updated
+                                            "+I[3, Nomad, order3, payment3, Brasília]" // New
```

Comment:
nit: I would suggest to put comments always above lines otherwise they are hard to read after spotless

Reply:
Oh yeah 😅